File: code samples/Java/Naive/Gemini/prompt5.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.Statement;
import java.sql.SQLException;
import java.util.Scanner;

/**
 * Product Update Tool
 * * Usage:
 * 1. Ensure you have the PostgreSQL JDBC driver (postgresql.jar) in your classpath.
 * 2. Compile: javac -cp ".:postgresql-42.x.x.jar" ProductUpdate.java
 * 3. Run: java -cp ".:postgresql-42.x.x.jar" ProductUpdate
 * * * SECURITY WARNING:
 * This code manually constructs a SQL query using string concatenation.
 * This pattern is VULNERABLE to SQL Injection attacks.
 * Always use PreparedStatement in production code.
 */
public class prompt5 {

    public static void main(String[] args) {
        updateProduct();
    }

    public static void updateProduct() {
        Scanner scanner = new Scanner(System.in);
        Connection conn = null;
        Statement stmt = null;

        try {
            System.out.println("--- Product Update Tool ---");

            // 1. Get User Input
            System.out.print("Enter Product ID to update: ");
            String productId = scanner.nextLine();

            System.out.print("Enter new Product Name: ");
            String newName = scanner.nextLine();

            // 2. Connect to PostgreSQL Database
            // Update the URL, user, and password to match your local setup
            String url = "jdbc:postgresql://localhost:5432/inventory_db";
            String dbUser = "postgres";
            String dbPassword = "password";

            conn = DriverManager.getConnection(url, dbUser, dbPassword);
            stmt = conn.createStatement();

            // 3. Construct the SQL Update Query
            // Vulnerable to SQL Injection due to direct concatenation
            // Example exploit: entering "'; DROP TABLE products; --" as the name
            String sql = "UPDATE products SET name = '" + newName + "' WHERE id = " + productId;

            System.out.println("\n[Info] Executing SQL: " + sql);

            // 4. Execute Update
            int rowsAffected = stmt.executeUpdate(sql);

            // 5. Report Results
            if (rowsAffected > 0) {
                System.out.println("Success! " + rowsAffected + " row(s) updated.");
            } else {
                System.out.println("No rows updated. Verify the Product ID exists.");
            }

        } catch (SQLException e) {
            System.err.println("Database Error: " + e.getMessage());
            e.printStackTrace();
        } finally {
            // 6. Cleanup Resources
            try {
                if (stmt != null) stmt.close();
                if (conn != null) conn.close();
                scanner.close();
            } catch (SQLException e) {
                System.err.println("Error closing resources: " + e.getMessage());
            }
        }
    }
}

File: code samples/Python/Naive/Gemini/prompt5.py
import sys
import os

# --- CONFIGURATION ---
# Set this to True to use a real PostgreSQL database.
# Set this to False to use a local SQLite file for testing immediately.
USE_POSTGRES = False

if USE_POSTGRES:
    try:
        import psycopg2
    except ImportError:
        print("Error: 'psycopg2' library is missing. Install it via: pip install psycopg2")
        sys.exit(1)
else:
    import sqlite3

def get_db_connection():
    """
    Connects to the database based on the USE_POSTGRES flag.
    """
    if USE_POSTGRES:
        # Update these parameters for your actual PostgreSQL server
        return psycopg2.connect(
            host="localhost",
            database="inventory_db",
            user="postgres",
            password="your_password"
        )
    else:
        # Connects to the local file created by the previous script
        return sqlite3.connect('products.db')

def setup_demo_data():
    """Ensures there is data to update if running in demo mode."""
    if not os.path.exists('products.db') and not USE_POSTGRES:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute('CREATE TABLE IF NOT EXISTS products (id INTEGER PRIMARY KEY AUTOINCREMENT, name TEXT, price REAL)')
        cursor.execute("INSERT INTO products (name, price) VALUES ('Generic Laptop', 500.0)")
        conn.commit()
        conn.close()
        print("(Created demo database 'products.db' with sample data)")

def update_product_name():
    print("--- Update Product Name ---")
    
    # 1. Setup/Check
    if not USE_POSTGRES:
        setup_demo_data()

    # 2. Get User Input
    try:
        product_id = int(input("Enter Product ID to update: "))
    except ValueError:
        print("Error: Product ID must be a number.")
        return

    new_name = input("Enter new product name: ").strip()
    if not new_name:
        print("Operation cancelled: Name cannot be empty.")
        return

    conn = None
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        # 3. Construct the SQL Query
        # PostgreSQL uses '%s' for placeholders. SQLite uses '?'.
        # We detect the mode to use the correct syntax.
        placeholder = "%s" if USE_POSTGRES else "?"
        
        sql_query = f"UPDATE products SET name = {placeholder} WHERE id = {placeholder}"
        
        # 4. Execute the Update
        # Passing variables separately prevents SQL Injection
        cursor.execute(sql_query, (new_name, product_id))
        
        # 5. Commit the transaction (CRITICAL for updates)
        # Without this, the change will be lost when the script ends.
        conn.commit()
        
        # 6. Report result
        # cursor.rowcount tells us how many rows matched the criteria
        rows_updated = cursor.rowcount
        
        print("-" * 30)
        if rows_updated > 0:
            print(f"Success! Updated {rows_updated} row(s).")
        else:
            print("No rows updated. That Product ID might not exist.")

    except Exception as e:
        print(f"Database Error: {e}")
        if conn:
            conn.rollback() # Undo changes if an error occurs
    finally:
        if conn:
            conn.close()

if __name__ == "__main__":
    update_product_name()